R/utils.R

```diff
@@ -536,51 +536,98 @@ get_mimetype <- function(filename) {
                             empty = "application/octet-stream"))
 }
 
-generate_css_dist_html <- function(href,
+generate_css_dist_html <- function(tagdata,
                                    local = FALSE,
                                    local_path = NULL,
                                    prefix = NULL,
                                    as_is = FALSE) {
+  attribs <- names(tagdata)
+  allowed_attribs <- c("as",
+                       "crossorigin",
+                       "disabled",
+                       "href",
+                       "hreflang",
+                       "importance",
+                       "integrity",
+                       "media",
+                       "referrerpolicy",
+                       "rel",
+                       "sizes",
+                       "title",
+                       "type",
+                       "methods",
+                       "prefetch",
+                       "target",
+                       "charset",
+                       "rev")
+  if (!all(attribs %in% allowed_attribs)) {
+      stop(sprintf("The following specified stylesheet attributes are invalid: ",
+                    paste0(setdiff(attribs, allowed_attribs), collapse=", ")), call. = FALSE)
+  }
   if (!(local)) {
-    if (grepl("^(?:http(s)?:\\/\\/)?[\\w.-]+(?:\\.[\\w\\.-]+)+[\\w\\-\\._~:/?#[\\]@!\\$&'\\(\\)\\*\\+,;=.]+$",
-        href,
-        perl=TRUE) || as_is) {
-      sprintf("<link href=\"%s\" rel=\"stylesheet\">", href)
+    if (any(grepl("^(?:http(s)?:\\/\\/)?[\\w.-]+(?:\\.[\\w\\.-]+)+[\\w\\-\\._~:/?#[\\]@!\\$&'\\(\\)\\*\\+,;=.]+$",
+        tagdata,
+        perl=TRUE)) || as_is) {
+            if (is.list(tagdata))
+                glue::glue('<link ', glue::glue_collapse(glue::glue('{attribs}="{tagdata}"'), sep=" "), ' rel="stylesheet">')
+            else
+                glue::glue('<link ', glue::glue('href="{tagdata}"'), ' rel="stylesheet">')
     }
     else
       stop(sprintf("Invalid URL supplied in external_stylesheets. Please check the syntax used for this parameter."), call. = FALSE)
   } else {
     # strip leading slash from href if present
+    if (is.list(tagdata))
+      href <- tagdata$src
+    else
+      href <- tagdata    
     href <- sub("^/", "", href)
     modified <- as.integer(file.mtime(local_path))
-    sprintf("<link href=\"%s%s?m=%s\" rel=\"stylesheet\">",
-            prefix,
-            href,
-            modified)
+    glue::glue('<link href="{prefix}{href}?m={modified}" rel="stylesheet">')
```

Comment:
Since at https://github.com/plotly/dashR/pull/226/files#diff-cbb7235afee10c3646161358a7655b0aR581 we handle the list case, I was expecting this to handle both the list and href-string cases like here https://github.com/plotly/dashR/pull/226/files#diff-cbb7235afee10c3646161358a7655b0aR571

Reply:
Excellent point, fixed in b2ff7a1.